File: python/with_pool.py
import time
import threading
from psycopg2 import pool as pg_pool
import config

# ---------------------------------------------------------------
# WITH Connection Pool  (psycopg2.pool.ThreadedConnectionPool)
#
# Pool is initialized ONCE in __init__.
#
# Two modes:
#
#   run_query()      — sequential, one request at a time
#                      Watch: same DB pid repeats = same connection reused
#
#   run_concurrent() — multiple threads fire simultaneously, each borrows
#                      its OWN connection from the pool at the same time.
#                      Watch: DIFFERENT pids in parallel = pool serving
#                      multiple connections concurrently.
#                      Also shows queuing when threads > pool size.
# ---------------------------------------------------------------

class WithPool:

    def __init__(self):
        print("[WITH POOL] Initializing psycopg2 ThreadedConnectionPool...")

        # ThreadedConnectionPool is thread-safe — safe to share across threads
        # minconn: opens these many connections immediately on startup
        # maxconn: hard ceiling — getconn() blocks if all are taken
        self._pool = pg_pool.ThreadedConnectionPool(
            minconn=config.POOL_MIN_CONN,
            maxconn=config.POOL_MAX_CONN,
            dsn=config.DSN
        )

        print(
            f"[WITH POOL] Pool ready — "
            f"min-conn: {config.POOL_MIN_CONN}, "
            f"max-conn: {config.POOL_MAX_CONN}"
        )
        print()

    # ------------------------------------------------------------------
    # Sequential mode
    # ------------------------------------------------------------------

    def run_query(self, request_number: int):
        self._run_query_internal(request_number, thread_id=1)
        time.sleep(0.1)

    # ------------------------------------------------------------------
    # Concurrent mode
    # ------------------------------------------------------------------

    def run_concurrent(self, total_threads: int, rounds: int):
        """
        Fires `total_threads` requests simultaneously each round.

        All threads are released at the EXACT same moment via threading.Barrier.

        What to watch:
          threads <= pool max  → all borrow instantly, DIFFERENT pids at same time
          threads >  pool max  → extras WAIT (borrow time spikes) = pool exhaustion
        """
        print(
            f"[CONCURRENT] Simulating {total_threads} simultaneous users | "
            f"Pool max-conn: {config.POOL_MAX_CONN}"
        )

        if total_threads > config.POOL_MAX_CONN:
            print(
                f"[CONCURRENT] *** {total_threads} threads > pool size {config.POOL_MAX_CONN}"
                f" → expect some threads to WAIT for a connection ***"
            )
        else:
            print(
                f"[CONCURRENT] {total_threads} threads <= pool size {config.POOL_MAX_CONN}"
                f" → all threads should get connections immediately"
            )
        print()

        request_counter = [0]           # shared counter (list so closure can mutate)
        counter_lock    = threading.Lock()

        for round_num in range(1, rounds + 1):
            print(f"--- Round {round_num} (all {total_threads} threads fire simultaneously) ---")

            # Barrier makes all threads wait until every thread has called wait()
            # then releases them ALL at the same instant — true simultaneous start
            start_barrier = threading.Barrier(total_threads)
            all_done      = threading.Event()
            done_count    = [0]
            done_lock     = threading.Lock()

            threads = []

            for t in range(total_threads):
                thread_id = t + 1

                with counter_lock:
                    request_counter[0] += 1
                    req_num = request_counter[0]

                def worker(req=req_num, tid=thread_id):
                    start_barrier.wait()        # hold until ALL threads are ready
                    self._run_query_internal(req, tid)
                    with done_lock:
                        done_count[0] += 1
                        if done_count[0] == total_threads:
                            all_done.set()

                thread = threading.Thread(target=worker, daemon=True)
                threads.append(thread)

            for t in threads:
                t.start()

            all_done.wait()     # block main thread until round is complete
            print()

    # ------------------------------------------------------------------
    # Internal — actual DB work, shared by both modes
    # ------------------------------------------------------------------

    def _run_query_internal(self, request_number: int, thread_id: int):
        start = time.monotonic()
        conn  = None

        try:
            # Borrow a connection from the pool — already open, near-zero cost
            # If pool is exhausted, this BLOCKS until one is returned
            conn = self._pool.getconn()
            conn_borrowed = time.monotonic()

            cursor = conn.cursor()

            # pg_sleep(0.05) holds the connection for 50ms so concurrent threads
            # are visibly active at the same time in the output
            cursor.execute("SELECT NOW(), pg_backend_pid(), pg_sleep(0.05)")
            row = cursor.fetchone()

            timestamp   = row[0]
            backend_pid = row[1]   # different pids = different pool connections used

            end = time.monotonic()

            borrow_ms = int((conn_borrowed - start)    * 1000)
            query_ms  = int((end - conn_borrowed)      * 1000)
            total_ms  = int((end - start)              * 1000)

            print(
                f"[CONCURRENT] Thread-{thread_id:<2} | "
                f"Req #{request_number:<3} | "
                f"Borrow: {borrow_ms:>4}ms | "
                f"Query: {query_ms:>3}ms | "
                f"Total: {total_ms:>4}ms | "
                f"DB pid: {backend_pid:<6}"
            )

            cursor.close()

        except Exception as e:
            end = time.monotonic()
            total_ms = int((end - start) * 1000)
            print(
                f"[CONCURRENT] Thread-{thread_id:<2} | "
                f"Req #{request_number:<3} | "
                f"FAILED after {total_ms}ms: {e}"
            )

        finally:
            if conn:
                # Return connection to pool — NOT closed, reused next time
                self._pool.putconn(conn)

    # ------------------------------------------------------------------

    def shutdown(self):
        print("\n[WITH POOL] Shutting down pool and closing all connections...")
        self._pool.closeall()

File: java/src/main/java/com/pooldemo/WithPool.java
package com.pooldemo;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * WITH Connection Pool (HikariCP)
 *
 * Pool is initialized ONCE in the constructor.
 *
 * Two modes available:
 *
 * runQuery()           — sequential, one request at a time (shows reuse via same pid)
 *
 * runConcurrent()      — multiple threads fire simultaneously, each borrows its OWN
 *                        connection from the pool at the same time.
 *                        Watch: DIFFERENT pids in parallel = pool serving multiple
 *                        connections concurrently from its pre-warmed set.
 *                        Also shows what happens when threads > pool size (queuing).
 */
public class WithPool {

    private final HikariDataSource dataSource;

    public WithPool() {
        System.out.println("[WITH POOL] Initializing HikariCP connection pool...");

        HikariConfig config = new HikariConfig();

        // Database coordinates
        config.setJdbcUrl(DBConfig.JDBC_URL);
        config.setUsername(DBConfig.USERNAME);
        config.setPassword(DBConfig.PASSWORD);

        // Pool sizing
        config.setMinimumIdle(DBConfig.POOL_MIN_IDLE);       // keep 5 connections warm always
        config.setMaximumPoolSize(DBConfig.POOL_MAX_SIZE);   // max 10 connections total

        // How long a request waits for a free connection before exception
        config.setConnectionTimeout(DBConfig.POOL_TIMEOUT_MS);

        // Kill connections idle more than 10 minutes
        config.setIdleTimeout(600_000);

        // Recycle any connection older than 30 minutes (avoids stale/zombie conns)
        config.setMaxLifetime(1_800_000);

        // Lightweight query used to verify a borrowed connection is still alive
        config.setConnectionTestQuery("SELECT 1");

        config.setPoolName("DemoPool");

        // *** This line opens min-idle connections to DB immediately ***
        this.dataSource = new HikariDataSource(config);

        System.out.printf("[WITH POOL] Pool ready — min-idle: %d, max-size: %d, timeout: %dms%n",
                DBConfig.POOL_MIN_IDLE, DBConfig.POOL_MAX_SIZE, DBConfig.POOL_TIMEOUT_MS);
        System.out.println();
    }

    /**
     * Fires `totalThreads` requests simultaneously using a thread pool.
     *
     * All threads start at the EXACT same moment (via CountDownLatch).
     * Each thread independently borrows a connection from the HikariCP pool.
     *
     * What to watch:
     *  - When totalThreads <= pool max-size  → all threads get a connection immediately
     *    (DIFFERENT DB pids running in parallel at the same time)
     *  - When totalThreads > pool max-size   → extra threads wait (Borrow time goes up)
     *    This is the pool exhaustion scenario — threads queue for a free connection.
     *
     * @param totalThreads  how many concurrent "users" to simulate
     * @param rounds        how many rounds (waves) of concurrent requests to run
     */
    public void runConcurrent(int totalThreads, int rounds) throws InterruptedException {

        System.out.printf("[CONCURRENT] Simulating %d simultaneous users | Pool max-size: %d%n",
                totalThreads, DBConfig.POOL_MAX_SIZE);

        if (totalThreads > DBConfig.POOL_MAX_SIZE) {
            System.out.printf("[CONCURRENT] *** %d threads > pool size %d → expect some threads to WAIT for a connection ***%n",
                    totalThreads, DBConfig.POOL_MAX_SIZE);
        } else {
            System.out.printf("[CONCURRENT] %d threads <= pool size %d → all threads should get connections immediately%n",
                    totalThreads, DBConfig.POOL_MAX_SIZE);
        }
        System.out.println();

        ExecutorService executor = Executors.newFixedThreadPool(totalThreads);
        AtomicInteger requestCounter = new AtomicInteger(0);

        for (int round = 1; round <= rounds; round++) {
            System.out.printf("--- Round %d (all %d threads fire simultaneously) ---%n", round, totalThreads);

            // Latch holds all threads until everyone is ready — true simultaneous start
            CountDownLatch startGun   = new CountDownLatch(1);
            CountDownLatch allDone    = new CountDownLatch(totalThreads);

            for (int t = 0; t < totalThreads; t++) {
                final int threadId = t + 1;
                final int reqNum   = requestCounter.incrementAndGet();

                executor.submit(() -> {
                    try {
                        startGun.await(); // all threads wait here until we fire the latch
                        runQueryInternal(reqNum, threadId);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                    } finally {
                        allDone.countDown();
                    }
                });
            }

            startGun.countDown();           // fire! all threads released at same instant
            allDone.await();                // wait for this round to finish
            System.out.println();
        }

        executor.shutdown();
        executor.awaitTermination(30, TimeUnit.SECONDS);
    }

    /**
     * Internal method used by both runQuery() and runConcurrent().
     * Actual DB work happens here.
     */
    private void runQueryInternal(int requestNumber, int threadId) {
        long start = System.currentTimeMillis();

        try (Connection conn = dataSource.getConnection();
             Statement stmt  = conn.createStatement()) {

            long connBorrowed = System.currentTimeMillis();

            // pg_sleep(0.05) makes the query hold the connection for 50ms
            // so we can clearly see multiple connections held simultaneously
            ResultSet rs = stmt.executeQuery(
                "SELECT current_timestamp, pg_backend_pid(), pg_sleep(0.05)"
            );

            if (rs.next()) {
                String timestamp = rs.getString(1);
                int backendPid   = rs.getInt(2);

                long end = System.currentTimeMillis();

                System.out.printf(
                    "[CONCURRENT] Thread-%-2d | Req #%-3d | Borrow: %4dms | Query: %3dms | Total: %4dms | DB pid: %-6d%n",
                    threadId,
                    requestNumber,
                    (connBorrowed - start),   // 0ms if connection available, >0ms if had to wait
                    (end - connBorrowed),
                    (end - start),
                    backendPid                // different pids = different connections used in parallel
                );
            }

        } catch (Exception e) {
            long end = System.currentTimeMillis();
            System.err.printf(
                "[CONCURRENT] Thread-%-2d | Req #%-3d | FAILED after %dms: %s%n",
                threadId, requestNumber, (end - start), e.getMessage()
            );
        }
        // Connection returned to pool here — available for next round immediately
    }

    // Keep the original sequential method as-is
    public void runQuery(int requestNumber) {
        runQueryInternal(requestNumber, 1);
        try { Thread.sleep(100); } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
    }

    public void shutdown() {
        System.out.println("[WITH POOL] Shutting down pool and closing all connections...");
        dataSource.close();
    }
}
